terraform/eval_state.go

```diff
@@ -221,15 +355,14 @@ func (n *EvalUndeposeState) Eval(ctx EvalContext) (interface{}, error) {
 		return nil, nil
 	}
 
-	// If we don't have any tainted, then we don't have anything to do
-	if len(rs.Tainted) == 0 {
+	// If we don't have any desposed resource, then we don't have anything to do
+	if len(rs.Deposed) == 0 {
 		return nil, nil
 	}
 
-	// Undepose to the tainted
-	idx := len(rs.Tainted) - 1
-	rs.Primary = rs.Tainted[idx]
-	rs.Tainted[idx] = nil
+	// Undepose
+	rs.Primary = rs.Deposed[len(rs.Deposed)-1]
+	rs.Deposed = rs.Deposed[:len(rs.Deposed)-1]
```

Comment:
Weird go thing here. We need to do this for GC reasons: `rs.Deposed[idx] = nil` (I did it for tainted before see above). The reason is because the reference is still in the slice, even though its no longer reachable, so the GC will never GC that value until `rs.Deposed` is also completely unreachable.

If you set the value to nil, then it removes the reference.